conan/tools/meson/toolchain.py

```diff
@@ -314,3 +342,45 @@ def generate(self):
         else:
             self._write_native_file()
         VCVars(self._conanfile).generate()
+
+
+class MesonAppleMixin(object):
+    def __init__(self, conanfile):
+        if not cross_building(conanfile):
+            return
+        if self._base_compiler != 'apple-clang':
+            return
+        if not is_apple_os(conanfile.settings.get_safe("os")):
+            return
+
+        xcrun = XCRun(conanfile.settings)
+        self.c = self.c or self._to_meson_value(xcrun.cc)
+        self.cpp = self.cpp or self._to_meson_value(xcrun.cxx)
+
+        arch = to_apple_arch(conanfile.settings.get_safe("arch"))
+        if arch:
+            self.extend_args(["-arch", arch])
+
+        os_version = conanfile.settings.get_safe("os.version")
+        if os_version:
+            flag = apple_min_version_flag(conanfile)
+            self.extend_args([flag])
+
+        if xcrun.sdk_path:
+            self.extend_args(["-isysroot", xcrun.sdk_path])
+            self.root = self._to_meson_value(xcrun.sdk_path)
+
+    def extend_args(self, args):
+        self.c_args.extend(args)
+        self.c_link_args.extend(args)
+        self.cpp_args.extend(args)
+        self.cpp_link_args.extend(args)
+        self.objc_args.extend(args)
+        self.objc_link_args.extend(args)
+        self.objcpp_args.extend(args)
+        self.objcpp_link_args.extend(args)
+
+
+class MesonToolchain(MesonBaseToolchain, MesonAppleMixin):
```

Comment:
Please, no multiple inheritance. And avoid inheritance as much as possible. Go with composition unless there is an extremely compelling case.

Reply:
I think it's a well-known mixin pattern?
/cc @jgsogo 
suggestion on how to to it better are welcomed

Reply:
@memsharded do you want me to implement `AppleSystemBlock` like it was done in `CMakeToolchain`?
https://github.com/conan-io/conan/blob/develop/conan/tools/cmake/toolchain.py#L280

Reply:
Yes, please, avoid inheritance and mixins unless it is strictly necessary. Composition scales better and is more flexible for user customization. Going with a "blocks" approach like the ``CMakeToolchain`` will be better, please go that way.

Reply:
I am looking at how to implement blocks similarly to how it was done in CMake toolchain.
there are several ideas, and I think I'll need some advice.
first off, there are two main things Apple block should define:
- compiler executables, `c` (known as `CC` in autotools world) and `cpp` (known as `CXX`)
- flags for compilers and linkers, like `c_args` (`CFLAGS`), `cpp_args` (`CXXFLAGS`), `link_args` (`LDFLAGS`), etc.

for executables, it should take the very least priority (only use when not explicitly defined), for variables, it should effectively append them to any existing.

then, unlike `CMake`, meson machine files are simple .ini files with very limited syntax. they are not turing-complete, and cannot contain conditions, loop, and other flow control instructions. there are very few operators supported, like `+` and `/` (https://mesonbuild.com/Machine-files.html#constants), but that's it.

on the other hand, there is an interesting feature that meson supports multiple cross-files overriding each other: https://mesonbuild.com/Machine-files.html#loading-multiple-machine-files

so, we may use something like `--cross-file conan_meson_cross_apple.ini --cross-file conan_meson_cross_main.ini`.

therefore, there are likely 3 major approaches on how to implement that:
1. use multiple cross-files overriding each other
2. use a single cross-file with several blocks
3. use a single cross-file with one block, but compose several contexts within the toolchain

1) is easy to implement, but it might be kinda misleading: too many files to deal with for the consumer, e.g. too long command lines, and easy to mess up with an order
2) might be tricky with a limited .ini syntax. e.g. I can't define the same value twice within the same machine file, I will get an error:
```
configparser.DuplicateOptionError
option 'buildtype' in section 'built-in options' already exists
```
and I can't check if it's already defined above, without using an external check in the code.

3) is also easy to implement, it will be just a matter of creating several context dictionaries and merging them according to our specific rules (executables override, flags append). but it's not really a blocks approach.

looking for the advice, thanks!